packages/firestore/src/remote/bloom_filter.ts

```diff
@@ -0,0 +1,91 @@
+/**
+ * @license
+ * Copyright 2022 Google LLC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+import { Md5, Integer } from '@firebase/webchannel-wrapper';
+
+import { newTextEncoder } from '../platform/serializer';
+import { debugAssert } from '../util/assert';
+
+export class BloomFilter {
+  readonly size: number;
+
+  constructor(
+    private readonly bitmap: Uint8Array,
+    padding: number,
+    private readonly hashCount: number
+  ) {
+    if (this.bitmap.length === 0) {
+      debugAssert(
+        padding === 0,
+        'A valid empty bloom filter should have 0 padding.'
+      );
+    }
+    this.size = this.bitmap.length * 8 - padding;
+    debugAssert(padding >= 0, 'Padding is negative.');
+    debugAssert(this.size >= 0, 'Bitmap size is negative.');
```

Comment:
It's impossible for the caller to specify a negative bitmap size; instead, assert on the padding argument and bitmap.length directly and assume that the size will be positive.

Reply:
Bump.